### PR TITLE
Enumerable module. 

### PR DESCRIPTION
**Pull request by nganifaith**
Added custom methods that work exactly like Ruby built-in methods

1. My_each for ruby each.
2. my_each_with_index. To return each value and its index.
3. my_select:  Returns values that meet a given condition.
4. my_all: Returns True if all elements meet a given condition false otherwise.
5. my_any: Returns True if any element meets a given condition False otherwise.
6. my_none: Return True if no element meets a given condition and False otherwise.
7. my_count: Return the number of occurrences of a given arg.

### CODE REVIEW

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_each` is not working as the original `each` method, also is not passing all test cases.

### `each` is an Enumerator that has these characteristics: 
- Returns an Enumerator if no block is given.
```
[1, 2, 3].each 
# #<Enumerator: [1, 2, 3]:each>
````
- It does not change or modify the original array/range

- it returns the array/range/hash itself if a block is given.
```
[1, 2, 3].each { |num| num }
# [1, 2, 3]
(1..3).each { |num| num * 2 }
# 1..3
````
- it applies the block for each element inside the array/range/hash
```
[1, 2, 3].each { |num| puts num }
# 1
# 2
# 3
# => [1, 2, 3]
```
[Enumerator#each](https://ruby-doc.org/core-2.7.2/Enumerator.html#method-i-each)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_each_with_index` is not working as the original `each_with_index` method, also is not passing all test cases.

### `each_with_index` is an Enumerable that has these characteristics: 
- Returns an Enumerator if no block is given.
```
[1, 2, 3].each_with_index
# #<Enumerator: [1, 2, 3]:each_with_index>
````
- It does not change or modify the original array/range

- it returns the array/range/hash itself if a block is given.
```
[1, 2, 3].each_with_index { |num, idx| num }
# [1, 2, 3]
(1..3).each { |num, idx| num * idx }
# 1..3
````
- it applies the block for each element inside the array/range/hash
```
[1, 2, 3].each_with_index { |num, idx| puts num, idx }
# 1
# 0
# 2
# 1
# 3
# 2
# => [1, 2, 3]
```
[Enumerator#each](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-each_with_index)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_all` is not working as the original `all?` method, also is not passing all test cases. 

### `all?` is an Enumerable that has these characteristics:
- Returns `true` if all elements are not false.
```
[1, 'string', true].all?
# true
```
- Returns `false` if it finds at least one element that is not true
```
[1, 'string', nil].all?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#all?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-all-3F)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_any` is not working as the original `any?` method, also is not passing all test cases. 

### `any?` is an Enumerable that has these characteristics:
- Returns `true` if at least one element is not false.
```
[1, false, nil].any?
# true
```
- Returns `false` if all elements are false.
```
[nil, false, nil].any?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#any?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-any-3F)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_none` is not working as the original `none?` method, also is not passing all test cases. 

### `none?` is an Enumerable that has these characteristics:
- Returns `true` if all elements are false.
```
[nil, nil, nil].none?
# true
```
- Returns `false` if it finds at least one element that is true
```
[1, 'string', nil].none?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#none?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-none-3F)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_count` is not working as the original `count` method, also is not passing all test cases. 

### `count` is an Enumerable that has these characteristics:
- Returns the count of elements inside an array/range/hash.
- When a block is given it returns the count of elements that match the condition
```
[1, 2, 3].count { |num| num > 2 }
# 1
```
[Enumerable#count](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-count)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_map` is not working as the original `map` method, also is not passing all test cases.

### `map` is an Enumerable that has these characteristics:
- Returns an Enumerator if no block is given.
```
[1, 2, 3].map
# #<Enumerator: [1, 2, 3]:map>
```
- Returns a **new array** with the results of running a block
```
[1, 2, 3].map { |num| num * 2 }
# [2, 4, 6]
```
- doesn't change/modify the original array/range/hash
- can accept a proc as an argument
```
['1', '2', '3'].map(&:to_i)
# [1, 2, 3]
```
[Enumerable#map](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-map)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_select` is not working as the original `select` method, also is not passing all test cases.

### `select` is an Enumerable that has these characteristics:
- Returns an Enumerator if no block is given.
```
[1, 2, 3].select
# #<Enumerator: [1, 2, 3]:select>
```
- Returns a **new array** with the results that match the condition in the block
```
[1, 2, 3].select { |num| num > 2 }
# [3]
```
- doesn't change/modify the original array/range/hash

[Enumerable#select](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-select)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] `my_inject` is not working as the original `inject` method, also is not passing all test cases.

### `inject`` is an Enumerable that has these characteristics:
- Returns a `LocalJumpError` when no block is given and no symbol is passed as an argument.
```
[1, 2, 3].inject
# LocalJumpError (no block given)
```

- it does not modify/change the original array/range/hash.

- it can accept an initial value and/or a symbol as an argument.
```
[1, 2, 3].inject(2, :*)
# 12
```

[Enumerable#inject](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-inject)

File: enumerables.rb
Comment (jubaan) [inline enumerables.rb]: - [ ] This is ok for you to have on your local version of the project, but make sure to remove when submitting the production version of it. [OPTIONAL]

**Review by jubaan — CHANGES_REQUESTED** [review]

**Comment by arikarim** [comment]
Thanks very much for the review it was really helpful, 
we are glad to tell you that we made these changes:
# changes the methods as you asked.
# added for functionality to them like return enumerator.
#because of the linter problem we added a line as a test case on line 117 so it does not cause problems.

thanks very much.

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: `my_all?` method is failing the following test cases:

```
- %w[ant bear cat].my_all?(/t/)  
- [1, 2i, 3.14].my_all?(Numeric) 
- [nil, true, 99].my_all? 
- [].my_all? 
- (0..5).my_all? 
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb]: `my_any?` method has failing test cases. Ensure that it passes all the following tests.

```
- %w[ant bear cat].my_any?(/d/)  
- [nil, true, 99].my_any?(Integer) 
- [nil, true, 99].my_any? 
- [].my_any?  
- (0..5).my_any? 
```

File: enumerables.rb
Comment (ghost) [inline enumerables.rb:72]: `my_none? ` method has failing test cases. Ensure all the following test cases are passing.

```
- %w{ant bear cat}.my_none?(/d/) 
- [1, 3.14, 42].my_none?(Float) 
- [nil].my_none?  
- [nil, false].my_none?
- [nil, false, true].my_none?
-  (0..4).my_none? 
```

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required 🛑

## Hello team 👋,

**Great job implementing the requested changes 👏. However, there are a few changes needed which you will find in the comments below.**

## Happy coding 🥂.

**You can reach me on slack (Rose Wanjohi).**

**Comment by arikarim** [comment]
Thanks very much for the review it was really helpful and we could detect the problems and fix the code.
please have a look, thanks very much.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've made laudable progress in this project.
However, some changes are required before it can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**
_**You're having 2 failed test cases overall. Let's check them out**_
For this code review, here's my test data:
```ruby
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9
array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) }
my_proc = proc { |num| num > 10 }
```
- [ ] `#my_map` should execute only the proc when both a block and a proc are given and should return a new array with the results of running the proc once for every element in enum
```ruby
- Test case: expect(array.my_map(my_proc) { |num| num < 10 }).to eq array.map(&my_proc)
- Expected: [false, false, false, false, false, false, false, false, false, false, false, false, false, false, fa...e, false, false, false, false, false, false, false, false, false, false, false, false, false, false]
- Got: [true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true...rue, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true, true]
```
- [ ] enumerables #my_inject raises a "LocalJumpError" when no block or argument is given
```ruby
- Test case: expect { array.my_inject }.to raise_error(LocalJumpError)
- Expected: LocalJumpError,
- Got: <TypeError: nil is not a symbol nor a string>
```



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Comment by arikarim** [comment]
Thanks for the review, we did as you asked and improved the functions, please ha a look.

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @nganifaith and @arikarim  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is now passing all of the tests successfully and it is really nice, congratulations on the good work you have done so far. :muscle:
:+1: I liked that you have tried to implemented all of the requested changes from the previous reviews, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [x]  Files not related to the project should not be committed and added to .gitignore (e.g. .DS_Store) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/gitignore.png) from Pathwright) :point_right:  Remove the `3.0` file from the root directory of your project
- [x] I liked your readme file, but you need to  fix typing mistakes and grammatical errors in it. Please use [Grammarly](https://app.grammarly.com/) to spot those errors.
- [x] The setup section of your readme file is for the `bubble_sort` project, please adjust it according to this project specs.
- [x] The first author should have added his/her name on linkedin profile link instead of writing the `linkedin` word there, just do it as the second author did
- [x] It’s better to delete commented piece of codes. If you need to review a previous version of your code, use git history ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/no_inline_comments.png) from Pathwright)
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Comment by nganifaith** [comment]
Thank you @descholar-ceo for your review, We have done the required changes please do take another look at it. 
Thank you

**Review by mke2111 — APPROVED** [review]
# **🥇 🥇  STATUS: APPROVED 🥇 🥇** 

Congratulations **@arikarim** and **@nganifaith**, I applaud you for the splendid job done on this project. Keep up the efforts.

You may merge to the main branch and remember to submit a project completion form.

### Cheers 👍🏽 👍🏽 

**Slack: Roy Mukuye**

**Pull request merged**
